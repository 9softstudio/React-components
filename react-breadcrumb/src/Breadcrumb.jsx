import React, { Component } from 'react';
import PropTypes from 'prop-types';
import BreadcrumbItem from './BreadcrumbItem';

export default class Breadcrumb extends Component {
    constructor(props) {
        super(props);
    }

    static propTypes = {
        items: PropTypes.arrayOf(PropTypes.shape({
            label: PropTypes.string,
            value: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
        })).isRequired,
        value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
        onChange: PropTypes.func
    }

    onChange(item) {
        if (this.props.value === item.value) {
            return;
        }

        this.props.onChange && this.props.onChange(item);
    }

    _renderItem() {
        const { value, items } = this.props;
        let breadcrumbItems = [];

        for (let i = 0; i < items.length; i++) {
            const item = items[i];
            const isActive = item.value === value;

            breadcrumbItems.push(
                <BreadcrumbItem key={`breadcrumb-item${i}`} isActive={isActive} label={item.label} onClick={this.onChange.bind(this, item)} />
            );

            if (isActive) {
                break;
            }
        }
        
        return breadcrumbItems;
    }

    render() {
        const items = this._renderItem();
        if (items.length === 1) {
            return null;
        }

        return (
            <div className='breadcrumb'>
                {items}
            </div>    
        );
    }
}