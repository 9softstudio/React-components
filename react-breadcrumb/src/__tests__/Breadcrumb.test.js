import React from 'react'
import Breadcrumb from '../Breadcrumb'
import renderer from 'react-test-renderer'

describe('Breadcrumb', () => {
    const items = [{ label: 'level 1', value: 1 }, { label: 'level 2', value: '2' }];

    describe('render', () => {
        it('active element is first item - render null', () => { 
            const props = { items, value: 1 };

            const breadcrumb = new Breadcrumb(props);

            const result = breadcrumb.render(); 
        
            expect(result).toBeNull(); 
        });

        it('active element is first item - component render correctly', () => {
            const props = { items, value: '2' };

            const component = renderer.create(<Breadcrumb {...props} />); 

            const result = component.toJSON(); 

            expect(result).toMatchSnapshot(); 
        });
    });
    
     describe('onChange', () => {
         it('new value is the same with the current - do nothing', () => {
            const onChangeHandler = jest.fn();
            const props = { items, value: '2', onChange: onChangeHandler };

            const component = renderer.create(<Breadcrumb {...props} />);

            const breadcrumb = component.toJSON();
            const currentActiveItem = breadcrumb.children[1];
            currentActiveItem.props.onClick();
        
            expect(onChangeHandler).not.toHaveBeenCalled(); 
        });

        it('new value is not the same with the current - call onChange function with correct parameter', () => {
            const onChangeHandler = jest.fn();
            const props = { items, value: '2', onChange: onChangeHandler };

            const component = renderer.create(<Breadcrumb {...props} />);

            const breadcrumb = component.toJSON();
            const newActiveItem = breadcrumb.children[0];
            newActiveItem.props.onClick();
        
            expect(onChangeHandler).toHaveBeenCalledWith(items[0]); 
        });
    });
});