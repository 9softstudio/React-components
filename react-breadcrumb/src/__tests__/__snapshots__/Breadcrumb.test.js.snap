// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb render active element is first item - component render correctly 1`] = `
<div
  className="breadcrumb"
>
  <a
    className="breadcrumb-item "
    href="javascript:;"
    onClick={[Function]}
  >
    level 1
  </a>
  <a
    className="breadcrumb-item active"
    href="javascript:;"
    onClick={[Function]}
  >
    level 2
  </a>
</div>
`;
